src/coreclr/src/pal/src/thread/process.cpp

```diff
@@ -2851,65 +2852,63 @@ CreateProcessModules(
 
     // NOTE: the module path can have spaces in the name
     // __TEXT                 0000000196220000-00000001965b4000 [ 3664K  2340K     0K     0K] r-x/rwx SM=COW          /Volumes/Builds/builds/devmain/rawproduct/debug/build/out/Applications/Microsoft Excel.app/Contents/SharedSupport/PowerQuery/libcoreclr.dylib
-    char *line = NULL;
-    size_t lineLen = 0;
-    int count = 0;
-    ssize_t read;
 
-    char vmmapCommand[100];
-    int chars = snprintf(vmmapCommand, sizeof(vmmapCommand), "/usr/bin/vmmap -interleaved %d -wide", dwProcessId);
-    _ASSERTE(chars > 0 && chars <= sizeof(vmmapCommand));
+    // NOTE: Sometimes vmmap hides full paths to some process modules (.dylibs in non-system folders), causing debugger not to work.
+    // __TEXT                 000000010d8bd000-000000010ddce000 [ 5188K  5188K     0K     0K] r-x/rwx SM=COW          /Users/USER/*/libcoreclr.dylib
+    // So now we get modules information by iterating over regions using proc_pidinfo().  See dotnet/runtime#42888.
+    int count = 0;
 
-    FILE *vmmapFile = popen(vmmapCommand, "r");
-    if (vmmapFile == NULL)
+    uint64_t addr = 0;
+    for (;;)
     {
-        goto exit;
-    }
+        struct proc_regionwithpathinfo rwpi;
+        int sz = proc_pidinfo(dwProcessId, PROC_PIDREGIONPATHINFO, addr, &rwpi, sizeof rwpi);
+        if (sz != sizeof rwpi)
+        {
+            if (sz == 0 && errno == EINVAL)
+                break; // ok
 
-    // Reading maps file line by line
-    while ((read = getline(&line, &lineLen, vmmapFile)) != -1)
-    {
-        void *startAddress, *endAddress;
-        char moduleName[PATH_MAX];
+            DestroyProcessModules(listHead);
+            listHead = NULL;
+            count = 0;
+            break; // unknown error
+        }
+
+        const char *moduleName = rwpi.prp_vip.vip_path;
 
-        if (sscanf_s(line, "__TEXT %p-%p [ %*[0-9K ]] %*[-/rwxsp] SM=%*[A-Z] %[^\n]", &startAddress, &endAddress, moduleName, _countof(moduleName)) == 3)
+        bool dup = false;
+        for (ProcessModules *entry = listHead; entry != NULL; entry = entry->Next)
         {
-            bool dup = false;
-            for (ProcessModules *entry = listHead; entry != NULL; entry = entry->Next)
+            if (strcmp(moduleName, entry->Name) == 0)
             {
-                if (strcmp(moduleName, entry->Name) == 0)
-                {
-                    dup = true;
-                    break;
-                }
+                dup = true;
+                break;
             }
+        }
 
-            if (!dup)
+        if (!dup)
+        {
+            int cbModuleName = strlen(moduleName) + 1;
+            ProcessModules *entry = (ProcessModules *)InternalMalloc(sizeof(ProcessModules) + cbModuleName);
```

Comment:
Does it need a pairing `InternalDelete(entry)`?

Reply:
Not here, it's already done in DestroyProcessModules().